src/tests/performance/prio_test.rs

```diff
@@ -0,0 +1,313 @@
+async fn simulate_normal_peer(
+    node_addr: SocketAddr,
+    socket: TcpSocket,
+    start_barrier: Arc<Barrier>,
+) {
+    let mut synth_node = SyntheticNodeBuilder::default()
+        .build()
+        .await
+        .expect("unable to build a synthetic node");
+
+    // Establish peer connection
+    synth_node
+        .connect_from(node_addr, socket)
+        .await
+        .expect("unable to connect to node");
+
+    // Wait for all peers to connect
+    start_barrier.wait().await;
+
+    for i in 0..REQUESTS {
+        let message = Payload::UniEnsBlockReq(UniEnsBlockReq {
+            data_type: UniEnsBlockReqType::BlockAndCert,
+            round_key: ROUND_KEY,
+            nonce: i as u64,
+        });
+
+        // Query transaction via peer protocol.
+        if !synth_node.is_connected(node_addr) {
+            break;
+        }
+
+        synth_node
+            .unicast(node_addr, message)
+            .expect("unable to send message");
+
+        let now = Instant::now();
+
+        // We can safely drop the result here because we don't care about it - if the message is
+        // received and it's our response we simply register it for histogram and break the loop.
+        // In every other case we simply move out and go to another request iteration.
+        timeout(RESPONSE_TIMEOUT, async {
+            loop {
+                let m = synth_node.recv_message().await;
+                if matches!(&m.1, Payload::TopicMsgResp(TopicMsgResp::UniEnsBlockRsp(rsp))
+                     if rsp.block.is_some() && rsp.block.as_ref().unwrap().round == ROUND_KEY && rsp.cert.is_some()) {
+                    metrics::histogram!(METRIC_LATENCY, duration_as_ms(now.elapsed()));
+                    break;
+                }
+            }
+        }).await;
+    }
+
+    synth_node.shut_down().await
+}
+
+async fn simulate_malicious_peer(
+    node_addr: SocketAddr,
+    socket: TcpSocket,
+    start_barrier: Arc<Barrier>,
+) {
+    let mut synth_node = SyntheticNodeBuilder::default()
+        .build()
+        .await
+        .expect("unable to build a synthetic node");
+
+    // Establish peer connection
+    synth_node
+        .connect_from(node_addr, socket)
+        .await
+        .expect("unable to connect to node");
+
+    // Wait for all peers to start
+    start_barrier.wait().await;
+
+    for _ in 0..REQUESTS {
+        // Send a MsgOfInterest message with all expected tags included.
+        let tags = HashSet::from([
+            Tag::ProposalPayload,
+            Tag::AgreementVote,
+            Tag::MsgOfInterest,
+            Tag::MsgDigestSkip,
+            Tag::NetPrioResponse,
+            Tag::Ping,
+            Tag::PingReply,
+            Tag::ProposalPayload,
+            Tag::StateProofSig,
+            Tag::TopicMsgResp,
+            Tag::Txn,
+            Tag::UniEnsBlockReq,
+            Tag::VoteBundle,
+        ]);
+        let message = Payload::MsgOfInterest(MsgOfInterest { tags });

```

Comment:
aha, let's move this outside of the loop and then use `clone()` operation before sending - because creating a new hashmap and then constructing a payload might be more expensive than simply copying it in every loop iteration

same comment for the function above